selection.go

```diff
@@ -0,0 +1,267 @@
+
+	msg, err := stun.Build(stun.ClassRequest, stun.MethodBinding, transactionID,
+		&stun.Username{Username: a.remoteUfrag + ":" + a.localUfrag},
+		&stun.IceControlled{TieBreaker: a.tieBreaker},
+		&stun.Priority{Priority: local.Priority()},
+		&stun.MessageIntegrity{
+			Key: []byte(a.remotePwd),
+		},
+		&stun.Fingerprint{},
+	)
+
+	if err != nil {
+		a.log.Error(err.Error())
+		return
+	}
+
+	a.log.Tracef("ping STUN from %s to %s\n", local.String(), remote.String())
+	if overflow := len(a.pendingBindingRequests) - maxPendingBindingRequests; overflow > 1 {
+		a.log.Debugf("Discarded %d pending binding requests, pendingBindingRequests is full", overflow)
+		a.pendingBindingRequests = a.pendingBindingRequests[overflow:]
+	}
+
+	a.pendingBindingRequests = append(a.pendingBindingRequests, bindingRequest{
+		transactionID: transactionID,
+		destination:   remote.addr(),
+	})
+	a.sendSTUN(msg, local, remote)
+}
+
+func (s *ControlledSelector) HandleSucessResponse(m *stun.Message, local, remote *Candidate, remoteAddr net.Addr) {
+	// TODO according to the standard we should specifically answer a failed nomination:
+	// https://tools.ietf.org/html/rfc8445#section-7.3.1.5
+	// If the controlled agent does not accept the request from the
+	// controlling agent, the controlled agent MUST reject the nomination
+	// request with an appropriate error code response (e.g., 400)
+	// [RFC5389].
+
+	ok, pendingRequest := s.agent.handleInboundBindingSuccess(m.TransactionID)
+	if !ok {
+		s.agent.log.Errorf("discard message from (%s), invalid TransactionID %s", remote, m.TransactionID)
+		return
+	}
+
+	transactionAddr := pendingRequest.destination
+
+	// Assert that NAT is not symmetric
+	// https://tools.ietf.org/html/rfc8445#section-7.2.5.2.1
+	if !addrEqual(transactionAddr, remoteAddr) {
+		s.agent.log.Debugf("discard message: transaction source and destination does not match expected(%s), actual(%s)", transactionAddr, remote)
+		return
+	}
+
+	s.agent.log.Tracef("inbound STUN (SuccessResponse) from %s to %s", remote.String(), local.String())
+	s.agent.addValidPair(local, remote)
+
+	if bytes.Equal(s.validateNominationTxID, m.TransactionID) {
+		s.agent.sendBindingSuccess(s.nominationTxID, local, remote)
+	}
+}
+
+func (s *ControlledSelector) HandleBindingRequest(m *stun.Message, local, remote *Candidate) {
+	_, useCandidate := m.GetOneAttribute(stun.AttrUseCandidate)
+
+	if useCandidate {
+		// https://tools.ietf.org/html/rfc8445#section-7.3.1.5
+		p := s.agent.findValidPair(local, remote)
+
+		if p != nil {
+			// If the state of this pair is Succeeded, it means that the check
+			// previously sent by this pair produced a successful response and
+			// generated a valid pair (Section 7.2.5.3.2).  The agent sets the
+			// nominated flag value of the valid pair to true.
+			s.agent.setSelectedPair(p)
+			s.agent.sendBindingSuccess(m.TransactionID, local, remote)
+		} else {
+			p = newCandidatePair(local, remote, false)
```

Comment:
Is this to support aggressive nomination? In the normal case we shouldn't get a `success` for a pair that doesn't exist right?

Reply:
mmm not sure why not, there is no guarantee AFAIC the pair choosen by the controlling side had been processed by the controlled side

Reply:
I don't believe so, we aren't just nominating a validPair with the best priority. We want to nominate a pair that has seen a full Request/Response done by both sides.

```
   L                        R
   -                        -
   STUN request ->             \  L's
             <- STUN response  /  check

              <- STUN request  \  R's
   STUN response ->            /  check

   STUN request + attribute -> \  L's
             <- STUN response  /  check

(Flight 0)
- Controlled does nothing
- Controlling pings all pairs (Should respond with success)

(Flight 1)
- Controlled pings every candidate pair it got a BindingRequest on (Should respond with success)
- Controlling does nothing

(Flight 2)
- Controlled is waiting for a useCandidate
- Controlling should send `useCandidate` to any pair it get a BindingRequest from (Since controlled and controlling both had a full Request/Response each way)
```

Reply:
but we don't know if the full request/response was done by both sides AFAIK that's why we use the attribute. Maybe I'm missing something. The controlling knows which candidates have been accepted because of the response, but doesn't know which responses have been received by the controlled (the best thing it could do is to track his own responses, but that doesn't mean it had been received by the controlled one).

This is the relevant part of the spec:

```
https://tools.ietf.org/html/rfc8445#section-7.3.1.5
   If the controlled agent receives a Binding request with the USE-
   CANDIDATE attribute set, and if the ICE agent accepts the request,
   the following action is based on the state of the pair computed in
   Section 7.3.1.4:

   o  If the state of this pair is Succeeded, it means that the check
      previously sent by this pair produced a successful response and
      generated a valid pair (Section 7.2.5.3.2).  The agent sets the
      nominated flag value of the valid pair to true.

   o  If the received Binding request triggered a new check to be
      enqueued in the triggered-check queue (Section 7.3.1.4), once the
      check is sent and if it generates a successful response, and
      generates a valid pair, the agent sets the nominated flag of the
      pair to true.  If the request fails (Section 7.2.5.2), the agent
      MUST remove the candidate pair from the valid list, set the
      candidate pair state to Failed, and set the checklist state to
      Failed.
```

Reply:
Aye that is all right, thanks for pulling the RFC keep trucking @hugoArregui 

I am so excited to land this, so many bad connections were probably caused by this bad ICE implementation :(

Reply:
I'll put the second item in the code (the first one is already there), I love when the comments point to the RFC, it makes it so easy to follow!